### PR TITLE
Issue 4909 - Milestone 3: Step 1 - 3D Calibration points

### LINKED ISSUE #4909: [2D MVP] Milestone 3: Step 1 - 3D calibration points
### Description
For calibrating the 2d drawing with the 3d drawing we need to find a way to transform 3d points into 2d points (and vice versa). 
- In this step we are creating a 3d vector in the viewer to be mapped to a 2d vector in next step. 
What we see here is a split view between the 3d and the 2d viewer.
- In the 3d side we have a limited version of the sidebar and a special version of the toolbar.
- Mesh selection should be disabled in this viewer.
- If this is a re-calibration we'll use the data from the previous calibration

<details>
  <summary> Sidebar</summary>
 
 Though in the mockup we can see different icons (**don't implement them**)  the  icons showing are
  **Groups**,  **Views** and **Tree**. So these are the cards that should be working on this 3d vector step. 
  If the sidebar code is too complicated we should aim to refactor it and make it simple.

</details>

<details>
  <summary> Tool bar</summary>

  The toolbar is somewhat simplified also,  consisting of  **Home**, **Perspective/Ortographic view**, **Turntable/Helicopter view**,  **sectioning**, **attribute data**. All of these should work as the are working currently in the viewer. We should create a new toolbar, reusing the buttons that we already have.

  There are also a new button  
###   (note: the vertical boundaries button shouldn't be there, it's wrong in figma):
  
  image
  
  - **Calibration vector** : when activated this button sets the viewer in a Vector mode. there also a couple of buttons that appears in the _subtoolbar_ to manage the  start and end points of the vector. 
  They are disabled while there is no vector defined.  When the user defines the start point the "clear start" button gets enabled and when the end point is the defined the "clear end" buttons is enabled. 
  Clicking on each of this buttons will delete its corresponding point from the vector
  
  Once its activated the user will define a vector in the 3d viewer. If the vector has been defined then the "Continue" button is enabled in the steps bar.
    
</details>

<details>
  <summary> 2d view </summary>
  
  It should only have *zoom in* and *zoom out* buttons

</details>

<details>
  <summary> Unity integration </summary>
  
  When clicking the "calibration vector" button  two things must happen
  
  1. Enable snapping by calling  _UnityUtil._**enableSnapping()** . 
  2. Set the calibration mode by calling _UnityUtil._**setCalibrationToolMode('Vector')** .
  3. If this is a recalibration use  _UnityUtil_.**setCalibrationToolVector(startPoint, endPoint)**  with **startPoint** and **endPoint** from the saved calibration. So the continue button will be available.

  
  The user is then abled to click on the model, create the start point and then click again and create the end point.
  
  When the user places either end of the vector, the viewer will call `UnityUtil.calibrationVectorChanged`. This method will be passed an object containing the start and end points of the vector, like so:
  
  ```
  {
      "start": [
          -0.18804311752319336,
          0.999995231628418,
          0.25510096549987793
      ],
      "end": []
  }
  ```
  The arrays will be the 3D positions of that end in Project coordinates, or empty, depending on whether that end has been specified yet.
  
  - For clearing the start point  _UnityUtil_.**setCalibrationToolVector(null, endPoint)**
  - For clearing the end point  _UnityUtil_.**setCalibrationToolVector(startPoint, null)**
</details>

<details>
  <summary> Calibration format  </summary>

```js
  {
    // this will most likely live in the drawing history
    calibration: {
        horizontal: {
            model: [[0,0,0], [1,1,1]],
            drawing: [[0,0], [1,1]]
        },
        verticalRange: [ 0, 10],
        // considering different container can be in different units we'd need to make sure we know how to scale it when viewing on a different model.
       units: "m"
     }
}

When showing the calibration take in consideration that the units might be different from the units in the container. Calculate to compensate if necessary.

```

</details>

### Goals
- [ ]  As a calibration user I want to be able to create a 3d vector for calibration

### Tasks
- [x] Make this component appear when in the "3d calibration points" step
- [x] Create/update sidebar with the allowed functionality for this step.
- [x] Create a toolbar for calibration
- [x] Update the viewer with the new unity functions
- [x] If this is a recalibration call _UnityUtil_.**setCalibrationToolVector(startPoint, endPoint)** with the previous calibration data 
- [x] Trigger UnityUtil.setCalibrationToolMode('Vector') when clicking in the calibration vector button and enable snapping
- [x] "Save" somewhere the created vector 
- [x] Enable/disable the continue button in the steps bar matching the vector state.
- [x] Add help box on top (look at design)

### Related Resources
[MockUp](https://www.figma.com/file/uTWnaYZ8IyoTn2i2xKtdBo/2D---3D?type=design&node-id=143-28437&mode=design&t=7bXLKF16hA7XqUeP-0)

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #4909

#### Description
The 3d calibration is not part of the calibration process.

#### Test cases
The vector logic is not entirely in the frontend (unity is only informed of any updates to re-render the arrow) and the buttons to delete either end have been removed. The mechanism is now the same as the 2d vector:
- if neither the start, nor the end of the vector exist, the next click sets the start
- if only the start exists, the next click will set the end
- if both exist, the next click will set the start and delete the end

If only the start of the vector exists, stopping 3d calibration (clicking on the respective toolbar button) will reset the vector to the drawing 3d vector value, i.e. if the drawing has a saved 3d vector, it will revert to that otherwise, it will be set to `null` for both start and end.

Once both the end and the start of the vector are set, it should be possible to proceed to the next step, otherwise, the continue button should be disabled.

Confirming the calibration should save the 3d vector in redux and the next calibration should read that data, rendering the vector.

##### PS
This is a pull request as Unity changes (from #4858) haven't been merged in yet

### CODE REVIEW

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx [hunk 1]
```diff
@@ -29,8 +32,14 @@ export interface CalibrationContextType {
 	isCalibrating: boolean;
 	origin: string;
 	setOrigin: (origin: string) => void;
+	isCalibrating3D: boolean,
+	setIsCalibrating3D: (isCalibrating3D: boolean) => void;
+	vector3D: Vector3D,
+	setVector3D: React.Dispatch<React.SetStateAction<Vector3D>>;
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx:38]: Does this need to be like this?

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationHeader/calibrationHeader.component.tsx [hunk 2]
```diff
@@ -29,9 +32,20 @@ const STEPS = [
 ];
 
 export const CalibrationHeader = () => {
-	const { step, setStep, isStepValid, origin } = useContext(CalibrationContext);
+	const { teamspace, project } = useParams();
+	const { step, setStep, isStepValid, origin, vector3D, drawingId } = useContext(CalibrationContext);
 	const isLastStep = step === 2;
 
+	const getIsStepValid = () => {
+		if (step === 0) return !!(vector3D.start && vector3D.end);
```
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationHeader/calibrationHeader.component.tsx:40]: Maybe this is for a later issue, but perhaps we should also check that the start and end point are not equal

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx [hunk 3]
```diff
@@ -41,19 +50,31 @@ const defaultValue: CalibrationContextType = {
...
 	const [origin, setOrigin] = useState(generatePath(DRAWINGS_ROUTE, { teamspace, project }));
 	const [isCalibrating] = useSearchParam('isCalibrating', Transformers.BOOLEAN);
-	const [arrow2D, setArrow2D] = useState({ start: null, end: null });
+	const [isCalibrating3D, setIsCalibrating3D] = useState(false);
+	const [vector3D, setVector3D] = useState<Vector3D>(EMPTY_VECTOR);
+	const [arrow2D, setArrow2D] = useState(EMPTY_VECTOR);
+	const [drawingId] = useSearchParam('drawingId');
+	const drawing = DrawingsHooksSelectors.selectDrawingById(drawingId);
+
+	const resetVector3D = () => setVector3D(drawing?.vector3D || EMPTY_VECTOR);
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx:77]: I dont think is necessary to have resetVector3d here. I would move that to the calibration3dHandler

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationStep/calibration3DHandler/calibration3DHandler.component.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		};
+	}, []);
+
+	useEffect(() => {
+		if (step !== 0) {
+			setIsCalibrating3D(false);
+		}
+	}, [step]);
+
+	useEffect(() => { resetVector3D(); }, [drawing]);
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationStep/calibration3DHandler/calibration3DHandler.component.tsx:69]: I don't know, I think it looks cooler 😝

```suggestion
	useEffect(resetVector3D, [drawing]);
```

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationStep/calibration3DHandler/calibration3DHandler.component.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+			UnityUtil.setCalibrationToolMode('None');
+			setIsCalibrating3D(false);
+		};
+	}, []);
+
+	useEffect(() => {
+		if (step !== 0) {
+			setIsCalibrating3D(false);
+		}
+	}, [step]);
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationStep/calibration3DHandler/calibration3DHandler.component.tsx:67]: if you revert the change in [/calibrationStep.component.tsx](https://github.com/3drepo/3drepo.io/pull/5031/files#diff-10a7bb727fdfd746c734b3eafc4d7648fac2ced9ad8e80bd709b4df3e0103e96)  with the switch, this is not necessary because it will take care of itself when unmounting

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationStep/calibrationStep.component.tsx [hunk 6]
```diff
@@ -31,9 +31,11 @@ export const CalibrationStep = () => {
...
 	}, [show2DViewer]);
 
-	switch (step) {
-		case 0: return <Calibration3DStep />;
-		case 1: return <Calibration2DStep />;
-		default: return <VerticalSpatialBoundariesStep />;
-	}
+	return (
+		<>
+			<Calibration3DHandler />
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationStep/calibrationStep.component.tsx:36]: read previous comment.

File: frontend/src/v5/ui/routes/viewer/toolbar/toolbar.component.tsx [hunk 7]
```diff
@@ -14,67 +14,25 @@
...
-					onClick={() => setBIMPanelVisibililty(!showBIMPanel)}
-					title={formatMessage({ id: 'viewer.toolbar.icon.attributeData', defaultMessage: 'Attribute Data' })}
-				/>
-			</MainToolbar>
-			<SectionToolbar />
-		</ToolbarContainer>
-	);
-};
+export const Toolbar = () => (
+	<ToolbarContainer>
```
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/viewer/toolbar/toolbar.component.tsx:26]: have to admit, this looks cleaner.

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx (hunk 1 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx:38]: I copied the typescript definition of `setState` so that I can call this function with both these syntaxes:
- `setState(newState)`
- `setState((oldState) => { /** code... */ ; return newState; })`

I try to simplify it though

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationHeader/calibrationHeader.component.tsx (hunk 2 above)
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationHeader/calibrationHeader.component.tsx:40]: Fair point. Then we should implement the same on the 2d vector as well!

File: frontend/src/v4/constants/viewerGui.ts [hunk 8]
```diff
@@ -114,8 +114,7 @@ export const getCalibrationViewerLeftPanels = () =>  [
 	VIEWER_PANELS.GROUPS,
 	VIEWER_PANELS.VIEWS,
 	VIEWER_PANELS.TREE,
-].filter((panel) =>
-(clientConfigService.sequencesEnabled || panel !== VIEWER_PANELS.SEQUENCES)).map(getPanelConfig);
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply frontend/src/v4/constants/viewerGui.ts:118]: nope, long gone!

File: frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx [hunk 9]
```diff
@@ -41,19 +50,31 @@ const defaultValue: CalibrationContextType = {
...
+	setVector3D: () => {},
+	resetVector3D: () => {},
+	arrow2D: EMPTY_VECTOR,
 	setArrow2D: () => {},
+	drawingId: '',
 };
 export const CalibrationContext = createContext(defaultValue);
 CalibrationContext.displayName = 'CalibrationContext';
 
 export const CalibrationContextComponent = ({ children }) => {
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/dashboard/projects/calibration/calibrationContext.tsx:63]: image

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/store/drawings/drawings.types.ts [hunk 10]
```diff
@@ -43,12 +44,22 @@ export interface MinimumDrawing {
 	drawingNumber: string;
 }
 
+export interface ICalibration {
```
[the remark this answers is not in the document]
Comment (sanmont3drepo) [reply frontend/src/v5/store/drawings/drawings.types.ts:47]: (use your discretion)

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

**Review by The-Daniel — APPROVED** [review]
